### PR TITLE
Add prepareHeader to graphqlrequestbasequery

### PR DESCRIPTION
**Pull request by rumurillo**
This PR adds an optional `prepareHeaders` function to the `graphqlRequestBaseQuery` that'd allow injecting headers on requests. The function provides a Headers object and receives the Redux's `getState` function to have access to the Redux store.

cc @phryneas

### CODE REVIEW

File: packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    | { client: GraphQLClient }
+  prepareHeaders?: (
+    headers: Headers,
+    api: Pick<GraphqlBaseQueryApi, 'getState'>
+  ) => MaybePromise<Headers>
+}
+
+export type BaseQueryFn<
+  Args = any,
+  Result = unknown,
```
Comment (phryneas) [inline packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts:29]: I'm not sure if making Result in any way configurable other than `unknown` makes any sense here - that would have to be a type of all possible results of all possible queries to send to the server. Do you have an example on how you would use that?

Comment (rumurillo) [reply packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts:29]: Hmm, not quite sure on that one as well. Actually, `BaseQueryFn` has (almost) the same signature as https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/baseQueryTypes.ts#L34. I added that new type with almost the same signature to be able to change the type definition for `api` param to the newly-added `GraphqlBaseQueryApi` interface:

````
export type BaseQueryFn<Args = any, Result = unknown, Error = unknown, DefinitionExtraOptions = {}, Meta = {}> = (
  args: Args,
  api: GraphqlBaseQueryApi,
  extraOptions: DefinitionExtraOptions,
) => MaybePromise<QueryReturnValue<Result, Error, Meta>>;
````

Comment (rumurillo) [reply packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts:29]: Hey, @phryneas -- hope you're doing alright.

Checking back on this one, do you have any other question/suggestion? Something I should be looking into?

Thanks!

Comment (phryneas) [reply packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts:29]: Sorry, I've got quite turbulent times right now - including a Covid infection. I'll try to get back to things next work, but no guarantees at the moment.

Comment (phryneas) [reply packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts:29]: Actually, now that I could take a closer look at it, that type should not be copied over at all. I'll push another commit in a few minutes, can you please check if that works as expected for you?

Comment (rumurillo) [reply packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts:29]: Yes, sure! I can test it out right away!

Comment (rumurillo) [reply packages/rtk-query-graphql-request-base-query/src/GraphqlBaseQueryTypes.ts]: Yes! That worked. I just had to pass in `requestHeaders` in `new Headers(stripUndefined({))` and it's all working great.

**Comment by juanjoseluisgarcia** [comment]
Can we give this a little bit of a push to get it merged? This feature is becoming vital for my project and we are wary to use something that is not on the main branch for production. 🙂

**Comment by phryneas** [comment]
@juanjoseluisgarcia I'll merge it but a new version will have to wait for a few days. I'd recommend using the CI build from this branch, as master has no CI builds 😝

**Comment by phryneas** [comment]
@matthewjberger tbh I've been waiting for anyone to test https://github.com/reduxjs/redux-toolkit/pull/2030 and give feedback that it is working as expected. Could you please do that?

**Comment by matthewjberger** [comment]
Absolutely, I'll take a look and see how I can help. 😄 I'll try and get up to speed on that PR and how to test it out

**Comment by phryneas** [comment]
I've gone ahead and released @rtk-query/graphql-request-base-query 2.0.0

**Comment by rumurillo** [comment]
I just installed `"@rtk-query/graphql-request-base-query": "^2.0.0"` and `"@reduxjs/toolkit": "^1.8.0"` and I'm getting an error related to the `isPlainObject` function:


> SyntaxError: Named export 'isPlainObject' not found. The requested module '@reduxjs/toolkit' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:
>
> import pkg from '@reduxjs/toolkit';
const {isPlainObject: e}from"@reduxjs/toolkit";import{GraphQLClient: t,ClientError: r}from"graphql-request";function n(){return n=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e},n.apply(this,arguments)}const a=a=>{const s="client"in a?a.client:new t(a.url),o="requestHeaders"in a?a.requestHeaders:{};return async({document:t,variables:i},{getState:c,endpoint:u,forced:p,type:l,signal:d,extra:f})=>{try{var m;const r=null!=(m=a.prepareHeaders)?m:e=>e,g=new Headers(function(t){if(!e(t))return t;const r=n({},t);for(const[e,t]of Object.entries(r))void 0===t&&delete r[e];return r}(o)),q=await r(g,{getState:c,endpoint:u,forced:p,type:l,extra:f});return{data:await s.request({document:t,variables:i,signal:d,requestHeaders:q}),meta:{}}}catch(e){if(e instanceof r){const{name:t,message:r,stack:n,request:a,response:s}=e;return{error:{name:t,message:r,stack:n},meta:{request:a,response:s}}}throw e}}};export{a: graphqlRequestBaseQuery} = pkg;

Any clue, @phryneas? 🤔

**Comment by phryneas** [comment]
(...and that's why I've been asking for feedback on that for weeks now...)

@rumurillo I didn't really change the way it was packaged - but of course that import is new and it seems your bundler is not using the esm build of RTK. Well, seems like I have to release a semi-breaking 2.0.1 now, removing the esm entry point :/

**Comment by phryneas** [comment]
Okay... 2.0.1 ist out without an `exports` field - please try again.

**Comment by rumurillo** [comment]
> (...and that's why I've been asking for feedback on that for weeks now...)
> 
> @rumurillo I didn't really change the way it was packaged - but of course that import is new and it seems your bundler is not using the esm build of RTK. Well, seems like I have to release a semi-breaking 2.0.1 now, removing the esm entry point :/

I'm sorry I couldn't get to test it before -- it's been a couple of hectic weeks!

I'll test right away!

**Comment by rumurillo** [comment]
> Okay... 2.0.1 ist out without an `exports` field - please try again.

Yup! That fixed it. Thanks, Lenz.

**Comment by rumurillo** [comment]
Well, that did fix the `esm entry point` issue; however, using the `graphqlRequestBaseQuery` now seems to reject all requests with a 404 error:

````
<!DOCTYPE html>
<html>
   <head>
      <style data-next-hide-fouc="true">body{display:none}</style>
      <noscript data-next-hide-fouc="true">
         <style>body{display:block}</style>
      </noscript>
      <script>(function(){ // Create a queue, but don't obliterate an existing one! var analytics = window.analytics = window.analytics || []; // If the real analytics.js is already on the page return. if (analytics.initialize) return; // If the snippet was invoked already show an error. if (analytics.invoked) { if (window.console && console.error) { console.error('Segment snippet included twice.'); } return; } // Invoked flag, to make sure the snippet // is never invoked twice. analytics.invoked = true; // A list of the methods in Analytics.js to stub. analytics.methods = [ 'trackSubmit', 'trackClick', 'trackLink', 'trackForm', 'pageview', 'identify', 'reset', 'group', 'track', 'ready', 'alias', 'debug', 'page', 'once', 'off', 'on', 'addSourceMiddleware', 'addIntegrationMiddleware', 'setAnonymousId', 'addDestinationMiddleware' ]; // Define a factory to create stubs. These are placeholders // for methods in Analytics.js so that you never have to wait // for it to load to actually record data. The `method` is // stored as the first argument, so we can replay the data. analytics.factory = function(method){ return function(){ var args = Array.prototype.slice.call(arguments); args.unshift(method); analytics.push(args); return analytics; }; }; // For each of our methods, generate a queueing stub. for (var i = 0; i < analytics.methods.length; i++) { var key = analytics.methods[i]; analytics[key] = analytics.factory(key); } // Define a method to load Analytics.js from our CDN, // and that will be sure to only ever load it once. analytics.load = function(key, options){ // Create an async script element based on your key. var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true; script.src = "https://cdn.segment.com/analytics.js/v1/" + key + "/analytics.min.js"; // Insert our script next to the first script element. var first = document.getElementsByTagName('script')[0]; first.parentNode.insertBefore(script, first); analytics._loadOptions = options; }; analytics._writeKey = 'bUYDvxPG7llQo7EeoyqHNGiVkbT4qnMA'; // Add a version to keep track of what's in the wild. analytics.SNIPPET_VERSION = '4.15.3'; // Load Analytics.js with your key, which will automatically // load the tools you've enabled for your account. Boosh! analytics.load("bUYDvxPG7llQo7EeoyqHNGiVkbT4qnMA"); // Make the first page call to load the integrations. If // you'd like to manually name or tag the page, edit or // move this call however you'd like. analytics.page(); })(); </script>
      <meta charSet="utf-8"/>
      <meta name="viewport" content="width=device-width"/>
      <meta name="next-head-count" content="2"/>
      <noscript data-n-css=""></noscript>
      <script defer="" nomodule="" src="/_next/static/chunks/polyfills.js?ts=1646441567525"></script><script src="/_next/static/chunks/webpack.js?ts=1646441567525" defer=""></script><script src="/_next/static/chunks/main.js?ts=1646441567525" defer=""></script><script src="/_next/static/chunks/pages/_app.js?ts=1646441567525" defer=""></script><script src="/_next/static/chunks/pages/_error.js?ts=1646441567525" defer=""></script><script src="/_next/static/development/_buildManifest.js?ts=1646441567525" defer=""></script><script src="/_next/static/development/_ssgManifest.js?ts=1646441567525" defer=""></script>
      <noscript id="__next_css__DO_NOT_USE__"></noscript>
   </head>
   <body>
      <div id="__next">
         <div class="splashScreen_splashScreenContainer__Hrb63">
            <div class="loader"></div>
         </div>
         <style data-emotion-css="19n335a-ToastContainer">.css-19n335a-ToastContainer{box-sizing:border-box;max-height:100%;max-width:100%;overflow:hidden;padding:8px;pointer-events:none;position:fixed;z-index:1000;top:0;right:0;}</style>
         <div class="react-toast-notifications__container css-19n335a-ToastContainer"></div>
      </div>
      <script src="/_next/static/chunks/react-refresh.js?ts=1646441567525"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{},"buildId":"development","isFallback":false,"gip":true,"scriptLoader":[]}</script>
   </body>
</html>
````

:/

**Comment by phryneas** [comment]
That is suboptimal. @rumurillo can you debug that one a bit? And make sure you updated graphql-request to the new major, the dependency bump is why we also went up a major.

**Comment by rumurillo** [comment]
> That is suboptimal. @rumurillo can you debug that one a bit? And make sure you updated graphql-request to the new major, the dependency bump is why we also went up a major.
> 
> It is working fine in our example so I wouldn't know where to start to debug this: https://codesandbox.io/s/github/reduxjs/redux-toolkit/tree/master/examples/query/react/graphql

Of course, yes! I'll test it later today!

**Comment by rumurillo** [comment]
> That is suboptimal. @rumurillo can you debug that one a bit? And make sure you updated graphql-request to the new major, the dependency bump is why we also went up a major.
> 
> It is working fine in our example so I wouldn't know where to start to debug this: https://codesandbox.io/s/github/reduxjs/redux-toolkit/tree/master/examples/query/react/graphql

I updated graphql-request to latest and it seems like I'm getting the same error. Not sure what's going on.

**Comment by phryneas** [comment]
@rumurillo  I did change some things after this PR, you might want to check that. Your PR moved some options into a nested config object and I moved them back to top level if I remember correctly.

**Comment by phryneas** [comment]
@antwash could you please open a new issue for this and provide a TS playground or CodeSandbox reproducing the issue?

**Pull request merged**